### PR TITLE
add inverse_transform #297 method to FPCA

### PR DESCRIPTION
**Pull request by Clej**
**Reference issue:**
https://github.com/GAA-UAM/scikit-fda/issues/297#issue-775429060

**What does it address ?**
The `inverse_transform` method projects the functional principal score (coefficient value w.r.t to eigenfunctions basis) of a (or set of) sample back to the input space. In other words `FPCA.inverse_transform(FPCA.transform())` should approximate the identity map.

**Empirical tests:**
I've only tested the `inverse_transform` method on synthetic datasets (both FDataGrid and FDataBasis) and assessed the results in terms of 'identity regression' i.e. with QQ-plots on the distribution of the residuals where residuals = X_input.value - X_recovered.value, where:

- X_recovered is computed as FPCA.inverse_transform(FPCA.transform(X_input)),
- value is .coefficients or .data_matrix attribute, depending on the input data format.

X_input is generated according to:
```javascript
cov = Exponential(length_scale=0.5)
n_grid = 100
n_samples = 50

fd = make_gaussian_process(
    start=0,
    stop=4,
    n_samples=n_samples,
    n_features=n_grid,
    mean=lambda t: np.power(t, 2) + 5,
    cov=cov,
)
```
Herebelow are the resulting QQ-plots (computed with `scipy.stats.probplot`) when `X_recovered' is computed with 3 principal components:
- When `fd' is let in FDataGrid format:
slope ~0.675, intercept ~ -3.e-17 and an R^2 ~ 0.9994. Note that the residuals are computed in terms of funtional data values.
qqplot-fdatagrid

- When `fd' is mapped to FDataBasis format with a 4-order Bspline basis with cardinality 50:
-slope ~ 1.08, intercept -5.e-17 and R^2 ~0.9996. Note that the residuals are computed in terms of the coefficients in the (here Bspline) basis.
qqplot-fdatabasis

I can enhance the PR and provide further emprical results, just tell me :)

### CODE REVIEW

**Comment by vnmabus** [comment]
Thank you for the PR! As a first step, please fix the formatting issues detected, and try to add a test case in test_fpca.py, to prevent future regressions.

I will try to review it this weekend, in order for it to be included in the next release.

@hzzhyj, as the original author of the `FPCA` class your feedback would also be greatly appreciated, if you have time to review it.

**Comment by hzzhyj** [comment]
> Thank you for the PR! As a first step, please fix the formatting issues detected, and try to add a test case in test_fpca.py, to prevent future regressions.
> 
> I will try to review it this weekend, in order for it to be included in the next release.
> 
> @hzzhyj, as the original author of the `FPCA` class your feedback would also be greatly appreciated, if you have time to review it.

I will try to review it over the weekend and early next week. Thanks for the PR too!

**Comment by Clej** [comment]
Okay! I'm hardly busy for the next two weeks, so I'll address the issues on the end of September.

File: skfda/preprocessing/dim_reduction/feature_extraction/_fpca.py [hunk 1]
```diff
@@ -326,6 +331,9 @@ def _fit_grid(
         # get the number of samples and the number of points of descretization
         n_samples, n_points_discretization = fd_data.shape
 
+        # necessary for inverse_transform
+        self.n_samples_fitted_ = n_samples
```
Comment (vnmabus) [inline skfda/preprocessing/dim_reduction/feature_extraction/_fpca.py:335]: Maybe it should be `n_samples_` like in scikit-learn.

**Comment by vnmabus** [comment]
Ok, I fixed  a bug in `transform` modified the code according to our style guide. I need you to check the "allow edits from maintainers" checkbox to add my modifications, pease.

**Comment by Clej** [comment]
Ok, I've checked "allow edits from maintainers", the checkbox is ticked!

**Review by vnmabus — APPROVED** [review]

**Comment by vnmabus** [comment]
Thank you @Clej for your first contribution to this repo! I hope you will continue contributing to open source! 🎉

**Comment by Clej** [comment]
My pleasure :) and thanks for your collaboration. For sure, I'll continue 👍 

Would it be relevant to create a short notebook on the utility of ``inverse_transform`` ? e.g. I frequently use it for outlier detection by comparing the reconstruction error of suspicious samples (test samples) w.r.t those of training samples.

**Comment by vnmabus** [comment]
Yeah, sure! Is that method commonly used in practice? I see several mentions of that for multivariate data, but not for functional data, and I do not see it implemented as a standalone outlier detection method in any library. In addition, I guess that you should ensure that there are no outliers in the training data in that case, or they will affect PCA estimation, right?

**Comment by Clej** [comment]
I think it is used in practice but not widely spread because it has two drawbacks: (i) requires a clean training set (ii) requires an error threshold. In general, an outlier detection method based on reconstruction error is likely used (and it makes more sense) with robust variants of FPCA, since the latters are specifically designed in a double perspective: (i) FPCA and (ii) outlier detection.

In my opinion, comparing ``x`` and ``fpca.inverse_transform(fpca.fit_transform(x))`` may also be useful to analyse and visualize the information loss induced by the truncation, and so detecting outliers from the reconstruction errors, with classical FPCA, is just one step in the whole analysis and not a state-of-the-art approach (or at least a baseline).
